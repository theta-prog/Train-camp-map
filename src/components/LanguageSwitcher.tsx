import { useState } from 'react'
import { useLocale } from 'next-intl'
import { useRouter, usePathname } from '@/i18n/navigation'

export default function LanguageSwitcher() {
  const router = useRouter()
  const pathname = usePathname()
  const locale = useLocale()
  const [isOpen, setIsOpen] = useState(false)

  const currentLocale = locale || 'ja'
  
  const handleLanguageChange = (newLocale: string) => {
    setIsOpen(false)
    router.push(pathname, { locale: newLocale as 'ja' | 'en' })
  }

  const currentLanguage = currentLocale === 'en' ? 'English' : 'æ—¥æœ¬èªž'
  const currentFlag = currentLocale === 'en' ? 'ðŸ‡ºðŸ‡¸' : 'ðŸ‡¯ðŸ‡µ'

  return (
    <div className="relative">
      <button
        onClick={() => setIsOpen(!isOpen)}
        className="flex items-center space-x-2 bg-white text-gray-700 px-4 py-2 rounded-lg shadow-md hover:shadow-lg transition-all duration-200 border border-gray-200"
      >
        <span className="text-lg">{currentFlag}</span>
        <span className="font-medium">{currentLanguage}</span>
        <svg
          className={`w-4 h-4 transition-transform duration-200 ${isOpen ? 'rotate-180' : ''}`}
          fill="none"
          stroke="currentColor"
          viewBox="0 0 24 24"
        >
          <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M19 9l-7 7-7-7" />
        </svg>
      </button>

      {isOpen && (
        <div className="absolute right-0 mt-2 w-48 bg-white rounded-lg shadow-lg border border-gray-200 z-50">
          <div className="py-1">
            <button
              onClick={() => handleLanguageChange('ja')}
              className={`w-full text-left px-4 py-3 hover:bg-gray-50 transition-colors flex items-center space-x-3 ${
                currentLocale === 'ja' ? 'bg-green-50 text-green-700' : 'text-gray-700'
              }`}
            >
              <span className="text-lg">ðŸ‡¯ðŸ‡µ</span>
              <span className="font-medium">æ—¥æœ¬èªž</span>
              {currentLocale === 'ja' && (
                <svg className="w-4 h-4 ml-auto text-green-600" fill="currentColor" viewBox="0 0 20 20">
                  <path fillRule="evenodd" d="M16.707 5.293a1 1 0 010 1.414l-8 8a1 1 0 01-1.414 0l-4-4a1 1 0 011.414-1.414L8 12.586l7.293-7.293a1 1 0 011.414 0z" clipRule="evenodd" />
                </svg>
              )}
            </button>
            <button
              onClick={() => handleLanguageChange('en')}
              className={`w-full text-left px-4 py-3 hover:bg-gray-50 transition-colors flex items-center space-x-3 ${
                currentLocale === 'en' ? 'bg-green-50 text-green-700' : 'text-gray-700'
              }`}
            >
              <span className="text-lg">ðŸ‡ºðŸ‡¸</span>
              <span className="font-medium">English</span>
              {currentLocale === 'en' && (
                <svg className="w-4 h-4 ml-auto text-green-600" fill="currentColor" viewBox="0 0 20 20">
                  <path fillRule="evenodd" d="M16.707 5.293a1 1 0 010 1.414l-8 8a1 1 0 01-1.414 0l-4-4a1 1 0 011.414-1.414L8 12.586l7.293-7.293a1 1 0 011.414 0z" clipRule="evenodd" />
                </svg>
              )}
            </button>
          </div>
        </div>
      )}

      {isOpen && (
        <div 
          className="fixed inset-0 z-40" 
          onClick={() => setIsOpen(false)}
        />
      )}
    </div>
  )
}
