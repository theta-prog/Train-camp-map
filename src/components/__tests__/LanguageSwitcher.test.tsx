import '@testing-library/jest-dom'
import { fireEvent, render, screen } from '@testing-library/react'
import LanguageSwitcher from '../LanguageSwitcher'

// モック next-intl
const mockUseLocale = jest.fn(() => 'ja')
jest.mock('next-intl', () => ({
  useLocale: () => mockUseLocale()
}))

// モック next/navigation
const mockPush = jest.fn()
const mockPathname = '/test'

jest.mock('@/i18n/navigation', () => ({
  useRouter: () => ({
    push: mockPush
  }),
  usePathname: () => mockPathname
}))

describe('LanguageSwitcher', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  // スナップショットテスト - 効率的な100%UIカバレッジ
  describe('スナップショットテスト', () => {
    it('日本語ロケール（閉じている状態）', () => {
      const { container } = render(<LanguageSwitcher />)
      expect(container.firstChild).toMatchSnapshot()
    })

    it('ドロップダウン開いている状態', () => {
      const { container } = render(<LanguageSwitcher />)
      
      // ドロップダウンを開く
      const button = screen.getByRole('button')
      fireEvent.click(button)
      
      expect(container.firstChild).toMatchSnapshot()
    })
  })

  it('基本的なレンダリングテスト', () => {
    render(<LanguageSwitcher />)
    
    // ボタンが存在することを確認
    expect(screen.getByRole('button')).toBeInTheDocument()
    
    // 日本語表記があることを確認
    expect(screen.getByText('日本語')).toBeInTheDocument()
    
    // 国旗があることを確認
    expect(screen.getByText('🇯🇵')).toBeInTheDocument()
  })

  it('ドロップダウンの開閉が機能する', () => {
    render(<LanguageSwitcher />)
    
    const button = screen.getByRole('button')
    
    // 初期状態では英語オプションは表示されない
    expect(screen.queryByText('English')).not.toBeInTheDocument()
    
    // ボタンをクリックしてドロップダウンを開く
    fireEvent.click(button)
    
    // 英語オプションが表示される
    expect(screen.getByText('English')).toBeInTheDocument()
  })

  it('言語変更時にルーター遷移が呼ばれる', () => {
    render(<LanguageSwitcher />)
    
    // ドロップダウンを開く
    const button = screen.getByRole('button')
    fireEvent.click(button)
    
    // 英語オプションをクリック
    const englishButton = screen.getByText('English')
    fireEvent.click(englishButton)
    
    // ルーター遷移が呼ばれることを確認
    expect(mockPush).toHaveBeenCalledWith('/test', { locale: 'en' })
  })

  it('背景クリックでドロップダウンが閉じる', () => {
    render(<LanguageSwitcher />)
    
    const button = screen.getByRole('button')
    fireEvent.click(button)
    
    // 英語オプションが表示される
    expect(screen.getByText('English')).toBeInTheDocument()
    
    // 背景をクリック
    const overlay = document.querySelector('.fixed.inset-0.z-40')
    if (overlay) {
      fireEvent.click(overlay)
    }
    
    // 英語オプションが非表示になる
    expect(screen.queryByText('English')).not.toBeInTheDocument()
  })

  it('英語ロケールでのレンダリングとチェックマーク非表示', () => {
    // 英語環境でテスト
    mockUseLocale.mockReturnValue('en')
    
    render(<LanguageSwitcher />)
    
    const button = screen.getByRole('button')
    fireEvent.click(button)
    
    // 英語オプションには選択状態のスタイルが適用される
    const englishElements = screen.getAllByText('English')
    expect(englishElements).toHaveLength(2)
    const englishOption = englishElements[1]?.closest('button') // ドロップダウン内の要素
    expect(englishOption).toHaveClass('bg-green-50', 'text-green-700')
    
    // 英語にはチェックマークが表示される
    expect(englishOption?.querySelector('svg')).toBeInTheDocument()
    
    // 日本語にはチェックマークが表示されない（条件分岐line 49）
    const japaneseButton = screen.getByText('日本語').closest('button')
    expect(japaneseButton?.querySelector('svg')).toBeNull()
  })

  it('ドロップダウン背景クリックでの状態変更', () => {
    render(<LanguageSwitcher />)
    
    const button = screen.getByRole('button')
    fireEvent.click(button)
    
    // ドロップダウンが開いている（2つのEnglishテキストが存在）
    const englishElements = screen.getAllByText('English')
    expect(englishElements).toHaveLength(2)
    
    // 背景クリックでsetIsOpenが呼ばれる
    const overlay = document.querySelector('.fixed.inset-0.z-40')
    if (overlay) {
      fireEvent.click(overlay)
    }
    
    // ドロップダウンが閉じる（Englishテキストが1つだけになる）
    const englishElementsAfterClose = screen.getAllByText('English')
    expect(englishElementsAfterClose).toHaveLength(1)
  })

  it('handleLanguageChangeでルーター遷移', () => {
    mockUseLocale.mockReturnValue('ja')
    
    render(<LanguageSwitcher />)
    
    const button = screen.getByRole('button')
    fireEvent.click(button)
    
    // 日本語ボタンをクリック（現在の言語と同じ）
    const japaneseElements = screen.getAllByText('日本語')
    expect(japaneseElements).toHaveLength(2)
    const japaneseButton = japaneseElements[1] // ドロップダウン内の要素
    if (japaneseButton) {
      fireEvent.click(japaneseButton)
    }
    
    // ルーター遷移が呼ばれることを確認
    expect(mockPush).toHaveBeenCalledWith('/test', { locale: 'ja' })
  })

  it('locale未定義時のフォールバック処理', () => {
    // localeがundefinedの場合をテスト
    (mockUseLocale as jest.Mock).mockReturnValue(undefined)
    
    render(<LanguageSwitcher />)
    
    const button = screen.getByRole('button')
    fireEvent.click(button)
    
    // デフォルトで日本語が選択されている
    const japaneseElements = screen.getAllByText('日本語')
    const japaneseOption = japaneseElements[1]?.closest('button')
    expect(japaneseOption).toHaveClass('bg-green-50', 'text-green-700')
    
    // 日本語にはチェックマークが表示される
    expect(japaneseOption?.querySelector('svg')).toBeInTheDocument()
  })
})
