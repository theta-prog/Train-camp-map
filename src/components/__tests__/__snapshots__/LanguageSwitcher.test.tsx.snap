// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`LanguageSwitcher スナップショットテスト ドロップダウン開いている状態 1`] = `
<div
  class="relative"
>
  <button
    class="flex items-center space-x-2 bg-white text-gray-700 px-4 py-2 rounded-lg shadow-md hover:shadow-lg transition-all duration-200 border border-gray-200"
  >
    <span
      class="text-lg"
    >
      🇯🇵
    </span>
    <span
      class="font-medium"
    >
      日本語
    </span>
    <svg
      class="w-4 h-4 transition-transform duration-200 rotate-180"
      fill="none"
      stroke="currentColor"
      viewBox="0 0 24 24"
    >
      <path
        d="M19 9l-7 7-7-7"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
      />
    </svg>
  </button>
  <div
    class="absolute right-0 mt-2 w-48 bg-white rounded-lg shadow-lg border border-gray-200 z-50"
  >
    <div
      class="py-1"
    >
      <button
        class="w-full text-left px-4 py-3 hover:bg-gray-50 transition-colors flex items-center space-x-3 bg-green-50 text-green-700"
      >
        <span
          class="text-lg"
        >
          🇯🇵
        </span>
        <span
          class="font-medium"
        >
          日本語
        </span>
        <svg
          class="w-4 h-4 ml-auto text-green-600"
          fill="currentColor"
          viewBox="0 0 20 20"
        >
          <path
            clip-rule="evenodd"
            d="M16.707 5.293a1 1 0 010 1.414l-8 8a1 1 0 01-1.414 0l-4-4a1 1 0 011.414-1.414L8 12.586l7.293-7.293a1 1 0 011.414 0z"
            fill-rule="evenodd"
          />
        </svg>
      </button>
      <button
        class="w-full text-left px-4 py-3 hover:bg-gray-50 transition-colors flex items-center space-x-3 text-gray-700"
      >
        <span
          class="text-lg"
        >
          🇺🇸
        </span>
        <span
          class="font-medium"
        >
          English
        </span>
      </button>
    </div>
  </div>
  <div
    class="fixed inset-0 z-40"
  />
</div>
`;

exports[`LanguageSwitcher スナップショットテスト 日本語ロケール（閉じている状態） 1`] = `
<div
  class="relative"
>
  <button
    class="flex items-center space-x-2 bg-white text-gray-700 px-4 py-2 rounded-lg shadow-md hover:shadow-lg transition-all duration-200 border border-gray-200"
  >
    <span
      class="text-lg"
    >
      🇯🇵
    </span>
    <span
      class="font-medium"
    >
      日本語
    </span>
    <svg
      class="w-4 h-4 transition-transform duration-200 "
      fill="none"
      stroke="currentColor"
      viewBox="0 0 24 24"
    >
      <path
        d="M19 9l-7 7-7-7"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
      />
    </svg>
  </button>
</div>
`;
